semgrep-core/bin/Main.ml

```diff
@@ -380,8 +382,20 @@ let parse_generic lang file =
   Naming_AST.resolve lang ast;
   Constant_propagation.propagate lang ast;
   (*e: [[Main_semgrep_core.parse_generic()]] resolve names in the AST *)
-  ast
+  Left ast
+ (* This is a bit subtle, but we now store in the cache whether we had
+  * a timeout on this file. Indeed, semgrep now calls semgrep-core
+  * per rule, and if one file timeout during parsing, it would timeout
+  * for each rule, but we don't want to wait each time 5sec for each rule.
+  * So here we store the exn in the cache, and below we reraise it
+  * after we got it back from the cache.
+  *)
+ with Timeout -> Right Timeout
```

Comment:
Should we store just timeouts or any parse error as well?

Reply:
yep, maybe better cache any exn.
I think only when we got a timeout things are really annoying performance wise because we
will do n times 5s, but it's true that maybe sometimes parsing takes only 4s but end up in a parse error.
I'll change the diff.